core/build.gradle

```diff
@@ -42,90 +20,106 @@ dependencies {
     [ 'netty-transport', 'netty-codec-http2', 'netty-resolver-dns' ].each {
         compile "io.netty:$it"
     }
-    compile "io.netty:netty-transport-native-epoll:${versionOf('netty')}:linux-x86_64"
+    compile 'io.netty:netty-transport-native-epoll:linux-x86_64'
     compile 'io.netty:netty-tcnative-boringssl-static'
     runtime 'org.javassist:javassist'
 
     // Reactive Streams
     compile 'org.reactivestreams:reactive-streams'
+}
 
-    // Internal dependencies which will be shaded by the 'internalDeps' task below.
-    internalDependency 'com.spotify:completable-futures'
-    internalDependency 'it.unimi.dsi:fastutil'
-    internalDependency 'org.reflections:reflections'
+tasks.shadedJar.baseName = "${tasks.jar.baseName}-untrimmed"
 
-    // Used by the 'relocateInternalDependencies', because ShadowTask doesn't let me specify a file.
-    trimmedInternalDependency files(internalDependenciesJarTrimmed)
-}
+task trimShadedJar(type: ProGuardTask,
+                   group: 'Build',
+                   description: 'Shrinks the shaded JAR by removing unused classes.') {
+    dependsOn tasks.shadedJar
 
-task trimInternalDependencies(type: ProGuardTask) {
-    injars files(configurations.internalDependency.collect()), filter: '!META-INF/**'
-    outjars internalDependenciesJarTrimmed
-    libraryjars files(configurations.runtime.collect() - configurations.internalDependency.collect())
+    def shadedFile = tasks.shadedJar.archivePath
+    def shadedAndTrimmedFile = file(shadedFile.path.replaceFirst('-untrimmed-', '-shaded-'))
+    injars shadedFile
+    outjars shadedAndTrimmedFile
+    libraryjars files(configurations.runtime.collect())
     libraryjars file("${System.getProperty('java.home')}/lib/rt.jar")
 
     dontoptimize
     dontobfuscate
     dontwarn // Ignore the harmless 'missing classes' warnings related with the optional dependencies.
 
     keepattributes 'Signature, InnerClasses, Annotation'
+    keep "class !${shadedPackage}.**,com.linecorp.armeria.** { *; }"
 
-    // CompletableFutures
-    keepclasses 'com.spotify.futures.CompletableFutures'
-
-    // FastUtil
-    keepinterfaces 'it.unimi.dsi.fastutil.ints.Int2ObjectMap',
-                   'it.unimi.dsi.fastutil.ints.Int2ObjectFunction'
-    keepclasses 'it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap',
-                'it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap'
-
-    // Reflections
-    keepclasses 'org.reflections.Configuration',
-                'org.reflections.Reflections',
-                'org.reflections.scanners.ResourcesScanner',
-                'org.reflections.util.ClasspathHelper',
-                'org.reflections.util.ConfigurationBuilder',
-                'org.reflections.util.FilterBuilder'
+    // Guava
+    keepclass "${shadedPackage}.guava.base.Converter"
```

Comment:
It looks like maintaining this list will be pretty annoying - we can't rely on proguard to automatically detect?